### PR TITLE
Use Jonathan's implementation for the description extractor

### PR DESCRIPTION
**Pull request by martijnvermaat**
Migrated from [GitLab !2](https://git.lumc.nl/mutalyzer/mutalyzer/merge_requests/2)

Here I did most of the work of replacing the old description extractor module by the new one (was temporarily called `describe_c.py`), but there are still some bugs to be solved before we can merge this.

1. Some unit tests crash in the description extractor code.
2. Some unit tests fail due to incorrect generated description.

@jfjlaros, could you have a look at this?

### CODE REVIEW

File: mutalyzer/describe.py [hunk 1]
```diff
@@ -1,5 +1,3 @@
-#!/usr/bin/python
-
 """
 Prototype of a module that can generate a HGVS description of the variant(s)
 leading from one sequence to an other.
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: Is it still a prototype of a module, or really a module?

File: mutalyzer/describe.py [hunk 2]
```diff
@@ -7,172 +5,42 @@
 @requires: Bio.Seq
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: Can be removed.

File: mutalyzer/describe.py [hunk 3]
```diff
@@ -7,172 +5,42 @@
...
 
-class LCS(object):
+def printpos(s, start, end, fill=0):
     """
-    Class that calculates a Longest Common Substring matrix once and provides
-    a function to find the LCS of a submatrix.
+    For debugging purposes.
     """
+    # TODO: See if this can partially replace or be merged with the
+    #       visualisation in the __mutate() function of mutator.py
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: Should be `_visualise()`.

File: mutalyzer/describe.py [hunk 4]
```diff
@@ -185,653 +53,344 @@ def makeFSTables(table_id):
...
-    @type peptide: unicode
-    @returns: All 2-mers of {peptide} in order of appearance.
-    @rtype: list(unicode)
+    :arg peptide: A peptide sequence.
+    :type peptide: unicode
+
+    :returns: All 2-mers of {peptide} in order of appearance.
+    :rtype: list(unicode)
     """
     return map(lambda x: peptide[x:x+2], range(len(peptide) - 1))
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: Prefered style is

    return [peptide[x:x+2] for x in range(len(peptide) - 1)]

or

    return [a + b for (a, b) in zip(peptide, peptide[1:])]

which are also faster.

File: mutalyzer/describe.py [hunk 5]
```diff
@@ -185,653 +53,344 @@ def makeFSTables(table_id):
...
+    :arg alternative_peptide: Observed peptide sequence.
+    :type alternative_peptide: unicode
+    :arg fs: Frame shift table.
+    :type fs: dict
+
+    :returns: Whether {alternative_peptide} is a frameshift of {peptide}.
+    :rtype: bool
     """
     # Todo: This is a temporary fix to prevent crashing on frameshift
     #     detection (I think bug #124).
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: Trac #124: https://humgenprojects.lumc.nl/trac/mutalyzer/ticket/124

File: mutalyzer/describe.py [hunk 6]
```diff
@@ -185,653 +53,344 @@ def makeFSTables(table_id):
...
+    :type s2: unicode
 
-            if ins_length == 1:
-                return [DescribeRawVar(DNA=False, start=s1_start,
-                    startAA=s1[s1_start - 1], type="dup", shift=shift)]
-            return [DescribeRawVar(DNA=False, start=s1_start - ins_length + 1,
-                startAA=s1[s1_start - ins_length], end=s1_end,
-                endAA=s1[s1_end - 1], type="dup", shift=shift)]
+    :returns: A list of RawVar objects, representing the allele.
+    :rtype: list(RawVar)
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: If I understand correctly, they are `models._RawVar` objects?

File: mutalyzer/describe.py [hunk 7]
```diff
@@ -185,653 +53,344 @@ def makeFSTables(table_id):
...
+
+    extracted = extractor.extract(s1.encode('utf-8'), len(s1),
+                                  s2.encode('utf-8'), len(s2), 0)
+    for variant in extracted.variants:
+       # print (variant.type, variant.reference_start,
+       #     variant.reference_end, variant.sample_start,
+       #     variant.sample_end, variant.transposition_start,
+       #     variant.transposition_end)
+       # print (variant.type & extractor.TRANSPOSITION_OPEN, variant.type &
+       #     extractor.TRANSPOSITION_CLOSE)
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: Incorrect indentation.

File: mutalyzer/describe.py [hunk 8]
```diff
@@ -185,653 +53,344 @@ def makeFSTables(table_id):
...
-    return default
-#protein_description
+    :arg s1: Sequence 1.
+    :type s1: unicode
+    :arg s2: Sequence 2.
+    :type s2: unicode
 
-def describe(original, mutated, DNA=True):
+    :returns: A list of RawVar objects, representing the allele.
+    :rtype: list(RawVar)
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: If I understand correctly, they are `models._RawVar` objects?

File: mutalyzer/entrypoints/mutalyzer.py [hunk 9]
```diff
@@ -16,6 +16,15 @@
...
 
 
+class MyEncoder(json.JSONEncoder):
+    def default(self, o):
+        json_object = o.__dict__
+        json_object.update({"hgvs": unicode(o), "weight": o.weight()})
+
+        return json_object
+    #default
+#MyEncoder
```
Comment (martijnvermaat) [inline mutalyzer/entrypoints/mutalyzer.py]: This seems like a bit of a weird trick. In any case, we should follow the pattern from the `json` documentation to call `JSONEncoder.default(self, o)` as a fallback.

https://docs.python.org/2/library/json.html#json.JSONEncoder.default

File: mutalyzer/extractor_loader.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+import sys
+
+import json
+
+import describe
+
+class MyEncoder(json.JSONEncoder):
+    def default(self, o):
+        return o.__dict__
```
Comment (martijnvermaat) [inline mutalyzer/extractor_loader.py]: See comment in `entrypoints.mutalyzer`.

File: mutalyzer/extractor_loader.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    extracted_allele = describe.describe_dna(ref, alt)
+
+    #print "Description Extractor Version " + describe.extractor.VERSION
+    print extracted_allele
+    #print "JSON: " + json.dumps({"reference_sequence": ref, "sample_sequence": alt, "allele_description": extracted_allele}, cls=MyEncoder)
+
+#main
+
+if __name__ == "__main__":
+    main()
```
Comment (martijnvermaat) [inline mutalyzer/extractor_loader.py]: Is this file needed? If so, it should be refactored into `entrypoints.mutalyzer`.

File: mutalyzer/models.py [hunk 12]
```diff
@@ -105,23 +105,81 @@ class RawVar(ComplexModel):
...
     term = Mandatory.Integer
     hgvs = Mandatory.Unicode
-    hgvsLength = Mandatory.Integer
+    hgvs_length = Mandatory.Integer
 #RawVar
 
 
+class _RawVar(ComplexModel):
+    """
+    Used in MutalyzerOutput data type.
```
Comment (martijnvermaat) [inline mutalyzer/models.py]: This description is incorrect. (Same for some others below.)

File: mutalyzer/test.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+from __future__ import unicode_literals
+
+import json
+
+import describe
+
+class MyEncoder(json.JSONEncoder):
+    def default(self, o):
+        return o.__dict__
```
Comment (martijnvermaat) [inline mutalyzer/test.py]: See comment in `entrypoints.mutalyzer`.

File: mutalyzer/test.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+    extracted_allele = describe.describe_dna(ref, alt)
+
+    print extracted_allele
+    print json.dumps({"reference_sequence": ref, "sample_sequence": alt,
+        "allele_description": extracted_allele}, cls=MyEncoder)
+#main
+
+if __name__ == "__main__":
+    main()
```
Comment (martijnvermaat) [inline mutalyzer/test.py]: Is this file needed? If so, it should be refactored into `entrypoints.mutalyzer`.

File: mutalyzer/variant.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
+"""
+"""
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: Could you add a description of the module?

File: mutalyzer/describe.py (hunk 1 above)
Comment (jfjlaros) [inline mutalyzer/describe.py]: Let's call it a module now.

File: mutalyzer/describe.py (hunk 4 above)
Comment (jfjlaros) [inline mutalyzer/describe.py]: *

This code is no longer used, it is now part of the extractor module.

File: mutalyzer/describe.py (hunk 5 above)
Comment (jfjlaros) [inline mutalyzer/describe.py]: See *.

File: mutalyzer/variant.py [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+
+from __future__ import unicode_literals
+
+from Bio.SeqUtils import seq3
+
+from extractor import extractor
+
+from mutalyzer import models
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: I don't like that we are using Spyne models here, they should be very specific to the webservices. If anything, we should try moving these models to the webservice code and remove all use of them elsewhere.

See the `models.py` docstring.

File: mutalyzer/variant.py [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+
+weights = {
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: Is better named `WEIGHTS` I think.

File: mutalyzer/variant.py [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+    def weight(self):
+        weight = sum(map(lambda x: x.weight(), self))
+
+        if len(self) > 1:
+            return weight + (len(self) + 1) * extractor.WEIGHT_SEPARATOR
+        return weight
+    #weight
+#HGVSList
+
+class Allele(HGVSList):
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: Could use a brief description.

This class name is also used in `models`, would be better to prevent this.

File: mutalyzer/variant.py [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+    pass
+
+class ISeqList(HGVSList):
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: Could use a brief description.

File: mutalyzer/variant.py [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+        if self.type == "ins":
+            return len(self.sequence) * extractor.WEIGHT_BASE
+
+        inverse_weight = weights["inv"] if self.reverse else 0
+        return (self.weight_position * 2 + extractor.WEIGHT_SEPARATOR +
+            inverse_weight)
+    #weight
+#ISeq
+
+class DNAVar(models.DNAVar):
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: This class name is also used in `models`, would be better to prevent this.

File: mutalyzer/extractor_loader.py (hunk 11 above)
Comment (jfjlaros) [inline mutalyzer/extractor_loader.py]: This was a script to make the JSON for the visualisation tool.

File: mutalyzer/variant.py [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+
+        weight = self.weight_position
+        if self.start != self.end:
+            weight += self.weight_position + extractor.WEIGHT_SEPARATOR
+
+        return weight + weights[self.type] + self.inserted.weight()
+    #weight
+#DNAVar
+
+class ProteinVar(models.ProteinVar):
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: This class name is also used in `models`, would be better to prevent this.

File: mutalyzer/variant.py [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+        :arg type: Variant type.
+        :type type: unicode
+        :arg deleted: Deleted part of the reference sequence.
+        :type deleted: unicode
+        :arg inserted: Inserted part.
+        :type inserted: object
+        :arg shift: Amount of freedom.
+        :type shift: int
+        :arg term:
+        :type term:
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: It's not necessary to enumerate all arguments here (the method signature is also shown in the API docs), so better remove the last two lines or add actual information.

File: mutalyzer/variant.py [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+        self.shift = shift
+        self.term = term
+    #__init__
+
+    def __unicode__(self):
+        """
+        Give the HGVS description of the raw variant stored in this class.
+
+        Note that this function relies on the absence of values to make the
+        correct description. Also see the comment in the class definition.
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: Which comment?

File: mutalyzer/website/views.py [hunk 24]
```diff
@@ -269,20 +269,18 @@ def name_checker():
...
-
-        if output.getIndexedOutput('oldprotein', 0):
-            prot_allele = describe.describe(
-                output.getIndexedOutput('oldprotein', 0),
-                output.getIndexedOutput('newprotein', 0, default=''),
-                DNA=False)
-            if prot_allele:
-                extractedProt = describe.alleleDescription(prot_allele)
+            extracted = unicode(allele) #describe.allele_description(allele)
+        if prot_allele:
```
Comment (martijnvermaat) [inline mutalyzer/website/views.py]: This is dead code (`prot_allele` is always `''`).

File: requirements.txt [hunk 25]
```diff
@@ -1,4 +1,5 @@
 -e git+https://github.com/LUMC/magic-python.git#egg=Magic_file_extensions
+-e git+https://git.lumc.nl/mutalyzer/extractor.git@3148fad6b081c36e9675494f4896ba60f14707a9#egg=extractor
```
Comment (martijnvermaat) [inline requirements.txt]: I specified the current master commit hash here, to prevent accidental breakeage of our servers.

This is fine for now, but ideally Jonathan would tag the extractor releases.

**Comment by martijnvermaat** [comment]
I had a very quick look over the changes and added some comments.

My main concern is with mixing our models with Spyne classes, I really don't think that's a good idea.

I ignored PEP8 related things (you could try the [Flake8](http://flake8.readthedocs.org/) tool for example).

File: mutalyzer/variant.py [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+weights = {
+    "subst": extractor.WEIGHT_SUBSTITUTION,
+    "del": extractor.WEIGHT_DELETION,
+    "ins": extractor.WEIGHT_INSERTION,
+    "dup": extractor.WEIGHT_INSERTION,
+    "inv": extractor.WEIGHT_INVERSION,
+    "delins": extractor.WEIGHT_DELETION_INSERTION
+}
+
+class HGVSList(list):
```
Comment (martijnvermaat) [inline mutalyzer/variant.py]: We should avoid subclassing the built-in `list` type. I think composition would make more sense here.

Comment (jfjlaros) [inline mutalyzer/variant.py]: Could you elaborate?

Comment (martijnvermaat) [inline mutalyzer/variant.py]: The current implementation might give some surprises, for example:

    >>> a = HGVSList([1,2,3])
    >>> type(a[1:])
    list

    >>> b = HGVSList([4,5,6])
    >>> type(a + b)
    list

This can of course be fixed by implementing the relevant magic methods, but it already hints that we might not really want to subclass `list`. None of the `list` methods will use any of the methods we override, leading to the unexpected behaviour as above and I think that shows `list` is not really meant to be subclassed.

Is our type really a `list` (in the Python sense)? I think it is better modelled as an object containing a `list`:

    class HGVSList(object):
        def __init__(self):
            self.elems = []

        @property
        def weight(self):
            return ...

It might still make sense to make this object iterable (yielding the list elements) for convenience.

So in summary, I think subclassing `list` might lead to confusion while it doesn't really gain us anything. Composition is the simpler and expected solution.

See also: [composition over inheritance](https://encrypted.google.com/search?hl=en&q=composition%20over%20inheritance)

Comment (jfjlaros) [inline mutalyzer/variant.py]: I read the chapter from Python the hard way.

The reason I chose to subclass `list` is because we need iteration, indexing and slicing. Furthermore, if we want to combine allele descriptions, we probably need a function like pop() as well. It seems to me that this reduces the amount of code.

The behaviour you describe is indeed troublesome, I'll give composition a try.

**Comment by jfjlaros** [comment]
I addressed most of the comments, with the following exceptions:
- The JSON issues.
- The entrypoint issues.
- The naming of the classes (overlap with models.py).

Most of these issues will become irrelevant when we move this code to the extractor package.

File: mutalyzer/describe.py [hunk 27]
```diff
@@ -1,837 +1,233 @@
...
-        return protein_description(M, s1, s2, lcp, s1_end, lcp, s2_end)
-    #if
-    M = LCS(s1, s2, lcp, s1_end, s2_end, DNA=True)
-    return DNA_description(M, s1, s2, lcp, s1_end, lcp, s2_end)
-#describe
+    description = Allele()
+
+    fs1, fs2 = make_fs_tables(1)
+    longest_fs_f = max(find_fs(s1, s2, fs1), find_fs(s1, s2, fs2))
+    longest_fs_r = max(find_fs(s2, s1, fs1), find_fs(s2, s1, fs2))
```
Comment (martijnvermaat) [inline mutalyzer/describe.py]: @jfjlaros Missing defintitions for `make_fs_tables` and `find_fs`.

Comment (jfjlaros) [inline mutalyzer/describe.py]: This code is now implemented in the description extractor, but we have not made a connection with Mutalyzer yet.

Of course I could clean this up, but this is currently unreachable code.

Comment (martijnvermaat) [inline mutalyzer/describe.py]: Ah yes, should have made this comment on the extractor repository.
